File: src/test/java/NonStandardAccountNumberTests.java
import com.github.pauldambra.moduluschecker.Account.BankAccount;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;

public class NonStandardAccountNumberTests {

    @Test
    public void CanCreateNatWestAccounts() {
        BankAccount account = BankAccount.Of("601123", "0123456789");

        assertThat(account.accountNumber, is(equalTo("23456789")));
    }

    @Test
    public void CanCreateHyphenatedNatWestAccounts() {
        BankAccount account = BankAccount.Of("602345", "01-23456789");

        assertThat(account.accountNumber, is(equalTo("23456789")));
    }

    @Test
    public void CanCreateCoopBankAccounts() {
        BankAccount account = BankAccount.Of("089286", "1234567890");

        assertThat(account.accountNumber, is(equalTo("12345678")));
    }

    @Test
    public void CanCreateSantanderAccounts() {
        BankAccount account = BankAccount.Of("123456", "123456789");

        assertThat(account.sortCode, is(equalTo("123451")));
        assertThat(account.accountNumber, is(equalTo("23456789")));
    }

    @Test
    public void CanCreateSevenDigitAccounts() {
        BankAccount account = BankAccount.Of("123456", "1234567");

        assertThat(account.sortCode, is(equalTo("123456")));
        assertThat(account.accountNumber, is(equalTo("01234567")));
    }

    @Test
    public void CanCreateSixDigitAccounts() {
        BankAccount account = BankAccount.Of("123456", "123456");

        assertThat(account.sortCode, is(equalTo("123456")));
        assertThat(account.accountNumber, is(equalTo("00123456")));
    }
}

File: src/test/java/NonStandardAccountNumberTests.kt


import com.github.pauldambra.moduluschecker.account.BankAccount
import com.natpryce.hamkrest.assertion.assert
import com.natpryce.hamkrest.equalTo
import org.junit.Test

class NonStandardAccountNumberTests {

    @Test
    fun canCreateNatWestAccounts() {
        val account = BankAccount("601123", "0123456789")

        assert.that(account.accountNumber, equalTo("23456789"))
    }

    @Test
    fun canCreateHyphenatedNatWestAccounts() {
        val account = BankAccount("602345", "01-23456789")

        assert.that(account.accountNumber, equalTo("23456789"))
    }

    @Test
    fun canCreateCoopBankAccounts() {
        val account = BankAccount("089286", "1234567890")

        assert.that(account.accountNumber, equalTo("12345678"))
    }

    @Test
    fun canCreateSantanderAccounts() {
        val account = BankAccount("123456", "123456789")

        assert.that(account.sortCode, equalTo("123451"))
        assert.that(account.accountNumber, equalTo("23456789"))
    }

    @Test
    fun canCreateSevenDigitAccounts() {
        val account = BankAccount("123456", "1234567")

        assert.that(account.sortCode, equalTo("123456"))
        assert.that(account.accountNumber, equalTo("01234567"))
    }

    @Test
    fun canCreateSixDigitAccounts() {
        val account = BankAccount("123456", "123456")

        assert.that(account.sortCode, equalTo("123456"))
        assert.that(account.accountNumber, equalTo("00123456"))
    }
}
